src/librustc_mir/shim.rs

```diff
@@ -873,34 +869,66 @@ pub fn build_adt_ctor<'a, 'gcx, 'tcx>(infcx: &infer::InferCtxt<'a, 'gcx, 'tcx>,
         scope: OUTERMOST_SOURCE_SCOPE
     };
 
-    let variant_no = if adt_def.is_enum() {
-        adt_def.variant_index_with_ctor_id(def_id)
+    let variant_index = if adt_def.is_enum() {
+        adt_def.variant_index_with_ctor_id(ctor_id)
     } else {
         VariantIdx::new(0)
     };
 
-    // return = ADT(arg0, arg1, ...); return
-    let start_block = BasicBlockData {
-        statements: vec![Statement {
+    // Generate the following MIR:
+    //
+    // (return as Variant).field0 = arg0;
+    // (return as Variant).field1 = arg1;
+    //
+    // discriminant(return) = variant_index;
+    // return;
+    //
+    // We only need downcasts if there's more than one variant.
```

Comment:
I thought you always needed a downcast if you had an `enum`, but maybe I'm wrong?